posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,388 @@
+API. [The Array API standard](https://data-apis.org/array-api/latest/index.html)
+solves this problem by standardizing functionality across most array libraries.
+NumPy and CuPy have already adopted the Array API standard; it can be accessed
+via the `__array_namespace__` mechanism in NumPy >=1.22.0 and CuPy >=10.0.0
+(both to be released in the coming month).
+
+The common architecture of SciPy,
+scikit-learn, and scikit-image is that they use NumPy for array operations, and
+critical-to-performance parts are written using compiled code extensions.
+Unfortunately, compiled extensions rely on NumPy's internal memory
+representation and are restricted to CPU computations. For these reasons, Python
+code with compiled extensions is not compatible with NumPy's dispatching
+capabilities; it would not know how to deal with a GPU array. It's also very
+challenging to rewrite the libraries to use pure NumPy without sacrificing
+performance. Therefore we need a dispatching mechanism (similar to NumPy's)
+for the API of modules in SciPy and other projects that rely on compiled extensions.
+
+Let's imagine some SciPy module that equally depends on compiled code and NumPy
+functions. Removing all forced conversions to NumPy arrays inside the module
+would open the "Dispatcher" path, but only a small part of SciPy
+functionality would work for generic arrays. In the diagram, the NumPy
+Dispatcher is an abstract mechanism, concretely it could be an
+`__array_function/ufunc__`-based one for older versions of NumPy or
+`__array_namespace__` - for NumPy 1.22+.
+
+<p align="center">
+    <img
+     alt="A schema using NumPy dispatcher."
+     src="/images/2021/11/scipy-numpy-dispatch.png">
+    <i><br>A SciPy module with NumPy "Dispatcher".</i>
+</p>
+
+<!-- ![numpy dispatch schema](https://drive.google.com/uc?export=view&id=1-PySRskT3-76r_KiiKGfkC485dT6uKoV) -->
+
+For SciPy and similar libraries, we need to follow NumPy's example and add our
+own dispatching layer for the API to allow users to develop generic code that
+works with arrays from different libraries. Then it would be possible to
+directly reroute to specific array implementations bypassing compiled code
+paths.
+
+<p align="center">
+    <img
+     alt="A schema using SciPy dispatcher."
+     src="/images/2021/11/scipy-ndimage-dispatcher.png">
+    <i><br>A SciPy module with its own "Dispatcher".</i>
+</p>
+
+<!-- ![scipy.ndimage dispatch schema](https://drive.google.com/uc?export=view&id=1-ZoVxenDufJuUwcONV_mFY0NFNY3wOCq) -->
+
+## Goals, wishes, and constraints
+
+Goal: separate the interface from the implementation and let the dispatching
+system help users and library authors write generic code that works with
+different kinds of arrays.
+
+The new dispatching system implemented for SciPy and scikits should:
+
+* make the API extendable and add the possibility to support alternative array
+  libraries, not just NumPy
```

Comment:
Another thing just came to mind. What should also be supported is "array duck types", i.e. things that are not subclasses of the base array object in some library, but are compatible with it and use that library's API and implementations. For example, objects that implement `__array_function__` but don't subclass `numpy.ndarray`, or implement `__torch_function__` but don't inherit from `torch.Tensor`.

That will work fine with `uarray` (because it just tries all registered backends and let's them say if they accept the object or not), but it doesn't work with any of the multiple dispatch implementations AFAIK.

Reply:
I think with any multiple dispatch implementation new unknown objects should be directed to the default implementation and then accepted or rejected. In types this means that the default implementation accepts `typing.Any` or `object`.
That works fine with `uarray` because it kind of overwrites the default implementation.

Yes, we should add this to the list of requirements.

Reply:
> I think with any multiple dispatch implementation new unknown objects should be directed to the default implementation and then accepted or rejected.

I disagree I think. That effectively means you're going to do the wrong thing for duck arrays (example, [pytorch/nestedtensor](https://github.com/pytorch/nestedtensor/blob/b050671e2c3618d0ff15f468922d9ea514b0c695/nestedtensor/nested/nested.py#L200)). It should either try the default implementation last (which is what `uarray` does), or it should be a principled type based dispatch, which means all unknown objects should raise immediately without dispatch happening. Otherwise it becomes a mess where numpy-likes are accepted and other-array-likes are accepted too but end up doing being converted to numpy arrays rather than being treated by the correct library.

Reply:
Added:
```
+* support "array duck types", i.e. custom array objects that are compatible
+  with a particular array library via (for example) `__array_function__` or
+  `__torch_function__`.
```